app/javascript/mastodon/features/timeline/index.js

```diff
@@ -0,0 +1,179 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import StatusListContainer from '../ui/containers/status_list_container';
+import Column from '../../components/column';
+import ColumnHeader from '../../components/column_header';
+import {
+  updateTimeline,
+  deleteFromTimelines,
+  connectTimeline,
+  disconnectTimeline,
+} from '../../actions/timelines';
+import { addColumn, removeColumn, moveColumn } from '../../actions/columns';
+import createStream from '../../stream';
+
+const mapStateToProps = (state, ownprops) => ({
+  streamingAPIBaseURL: state.getIn(['meta', 'streaming_api_base_url']),
+  accessToken: state.getIn(['meta', 'access_token']),
+  hasUnread: state.getIn(['timelines', ownprops.timelineId, 'unread']) > 0,
+});
+
+@connect(mapStateToProps)
+export default class Timeline extends React.PureComponent {
+
+  static propTypes = {
+    dispatch: PropTypes.func.isRequired,
+    streamingAPIBaseURL: PropTypes.string.isRequired,
+    accessToken: PropTypes.string.isRequired,
+    expand: PropTypes.func.isRequired,
+    refresh: PropTypes.func,
+    streamId: PropTypes.string,
+    hasUnread: PropTypes.bool,
+    columnName: PropTypes.string.isRequired,
+    columnProps: PropTypes.object,
+    columnId: PropTypes.string,
+    multiColumn: PropTypes.bool,
+    emptyMessage: PropTypes.oneOfType([
+      PropTypes.element,
+      PropTypes.string,
+    ]),
+    icon: PropTypes.string.isRequired,
+    title: PropTypes.string.isRequired,
+    settings: PropTypes.element,
+    scrollName: PropTypes.string.isRequired,
+    timelineId: PropTypes.string.isRequired,
+  };
```

Comment:
at 19 lines, this is the second longest propTypes declaration in the entire code base. that's what makes me worried this is a bad refactor—if we have to do this much specialization now, what happens when we want to make things more complicated? Huge, clear cut violation of the open/close principle.